creme/billing/tests/test_line.py

```diff
@@ -1186,6 +1189,210 @@ def test_multi_save_lines07(self):
+            'product_line_formset-3-unit':            'month',
+        })
+
+        response = self.client.post(
+            self._build_msave_url(invoice),
+            data={
+                lines[0].entity_type_id: json_dump(data)
+            },
+        )
+
+        self.assertNoFormError(response)
+
+        self.assertEqual([
+            (1, 'Product 1'),
+            (2, 'Product 2'),
+            (3, 'Product 3'),
+            (4, 'New on the fly product'),
+        ], [
+            (line.order, line.on_the_fly_item)
+            for line in ProductLine.objects.order_by('order')
+        ])
+
+    @parameterized.expand([
+        ('Product 1', 3, [
+            (1, 'Product 2'),
+            (2, 'Product 3'),
+            (3, 'Product 1'),
+            (0, 'Service 1'),
+            (0, 'Service 2'),
+        ]),
+        ('Product 3', 1, [
+            (1, 'Product 3'),
+            (2, 'Product 1'),
+            (3, 'Product 2'),
+            (0, 'Service 1'),
+            (0, 'Service 2'),
+        ]),
+        ('Product 1', 2, [
+            (1, 'Product 2'),
+            (2, 'Product 1'),
+            (3, 'Product 3'),
+            (0, 'Service 1'),
+            (0, 'Service 2'),
+        ]),
+        ('Service 1', 3, [
+            (0, 'Product 1'),
+            (0, 'Product 2'),
+            (0, 'Product 3'),
+            (1, 'Service 2'),
+            (2, 'Service 1'),
+        ]),
+    ])
+    @skipIfCustomProductLine
+    def test_reorder_lines(self, target_name, next_order, expected):
+        user = self.login()
+        invoice = self.create_invoice_n_orgas('Invoice001', discount=0)[0]
+        default_vat = Vat.objects.default()
+
+        lines = {
+            f'Product {order}': ProductLine.objects.create(
+                user=user, related_document=invoice, vat_value=default_vat,
+                unit_price=Decimal('10'), on_the_fly_item=f'Product {order}'
+            ) for order in range(1, 4)
+        }
+
+        lines.update({
+            f'Service {order}': ServiceLine.objects.create(
+                user=user, related_document=invoice, vat_value=default_vat,
+                unit_price=Decimal('10'), on_the_fly_item=f'Service {order}'
+            ) for order in range(1, 3)
+        })
+
+        url = reverse('billing__reorder_line', args=(invoice.pk, lines[target_name].pk))
+
+        with patch('creme.billing.models.Invoice.save') as fake_invoice_save:
+            response = self.client.post(url, data={'target': next_order})
+            self.assertEqual(response.status_code, 200)
+
+        invoice.refresh_from_db()
+        self.assertEqual(
+            [
+                (line.order, line.on_the_fly_item)
+                for line in invoice.get_lines(ProductLine).order_by('order')
+            ] + [
+                (line.order, line.on_the_fly_item)
+                for line in invoice.get_lines(ServiceLine).order_by('order')
+            ],
```

Comment:
itertools.chain() or list-comprehension with a loop `for model in (ProductLine, ServiceLine)`